lms/djangoapps/email_marketing/signals.py

```diff
@@ -54,41 +55,33 @@ def add_email_marketing_cookies(sender, response=None, user=None,
         if sailthru_content:
             post_parms['cookies'] = {'anonymous_interest': sailthru_content}
 
-    try:
-        sailthru_client = SailthruClient(email_config.sailthru_key, email_config.sailthru_secret)
-        log.info(
-            'Sending to Sailthru the user interest cookie [%s] for user [%s]',
-            post_parms.get('cookies', ''),
-            user.email
-        )
-        time_before_call = datetime.datetime.now()
+    time_before_call = datetime.datetime.now()
+    sailthru_response = get_email_cookies_via_sailthru.delay(user.email, post_parms)
 
-        sailthru_response = \
-            sailthru_client.api_post("user", post_parms)
+    try:
+        # synchronous call to celery task with a timeout limit
+        sailthru_response.get(timeout=email_config.user_registration_cookie_timeout_delay,
+                              propagate=True)
+        cookie = sailthru_response.result
 
+    except TimeoutError as exc:
+        log.error("Timeout error while attempting to obtain cookie from Sailthru: %s", unicode(exc))
```

Comment:
> That said, it's hard/impossible to do what you're attempting all in one process - something has to keep track of the timeout and interrupt if it expires.

Yes need a separate thread for this or some built in mechanism like signals to track timeout. I guess .get() call on result does this internally.

> How/where are you doing this simulation? If it locally on your devstack, you'll need to use multiple processes with --setting=devstack_with_worker, otherwise the initial delay gets converted to a synchronous call, run in-process (Because envs.common.py envs.devstack.py sets CELERY_ALWAYS_EAGER).

No iam not using any devstack instance. To test delay/get sequence, i setup a local rabbitmq/celery instance, with a task to only check for timeout thing. 
Initially i used only single/main thread to push task to rabbitmq/celery. The caller code was blocking because of get() usage and timeout error was correctly raising in case of task timeout. To simulate prod multiuser behaviour locally i used threads for this. Here when i don't use any kind of locks/synchronization, all task gives me timeout error. I guess there is some shared code or something is not thread safe. But i guess on prod, each django caller would be a separate process, so we would not be hitting this case. 
Iam using docker for development, so i dont know for now how to pass --setting=devstack_with_worker for server startup. Secondly tested this code in a sandbox and it was working fine. In logs i could see celery task, so i assume the task was scheduled async because of .delay() and blocked on .get(), waiting for the result. Iam still trying by someway to generate a timeout exception to test my code changes.